src/nunavut/lang/cpp/templates/_composite_type.j2

```diff
@@ -28,44 +28,58 @@
-    /// version of the data type; or, on other words, it is the the maximum possible size of received objects of this type.
-    /// The size is specified in bytes (rather than bits) because by definition, extent is an integer number of bytes long.
-    /// When allocating a deserialization (RX) buffer for this data type, it should be at least extent bytes large.
-    /// When allocating a serialization (TX) buffer, it is safe to use the size of the largest serialized representation
-    /// instead of the extent because it provides a tighter bound of the object size; it is safe because the concrete type
-    /// is always known during serialization (unlike deserialization). If not sure, use extent everywhere.
-
-    static constexpr {{ typename_unsigned_length }} EXTENT_BYTES                    = {#- -#}
-        {{ composite_type.extent // 8 }}UL;
-    static constexpr {{ typename_unsigned_length }} SERIALIZATION_BUFFER_SIZE_BYTES = {#- -#}
-        {{ composite_type.inner_type.extent // 8 }}UL;
-    static_assert(EXTENT_BYTES >= SERIALIZATION_BUFFER_SIZE_BYTES, "Internal constraint violation");
-    static_assert({# -#}
-        EXTENT_BYTES < (std::numeric_limits<{{ typename_unsigned_bit_length }}>::max() /8U), {# -#}
-        "This message is too large to be handled by current types!");
+        /// Extent is the minimum amount of memory required to hold any serialized representation of any compatible
+        /// version of the data type; or, on other words, it is the the maximum possible size of received objects of this type.
+        /// The size is specified in bytes (rather than bits) because by definition, extent is an integer number of bytes long.
+        /// When allocating a deserialization (RX) buffer for this data type, it should be at least extent bytes large.
+        /// When allocating a serialization (TX) buffer, it is safe to use the size of the largest serialized representation
+        /// instead of the extent because it provides a tighter bound of the object size; it is safe because the concrete type
+        /// is always known during serialization (unlike deserialization). If not sure, use extent everywhere.
+        static constexpr {{ typename_unsigned_length }} ExtentBytes                  = {# -#}
+            {{ composite_type.extent // 8 }}UL;
+        static constexpr {{ typename_unsigned_length }} SerializationBufferSizeBytes = {# -#}
+            {{ composite_type.inner_type.extent // 8 }}UL;
+        static_assert(ExtentBytes >= SerializationBufferSizeBytes, "Internal constraint violation");
+        static_assert({# -#}
+            ExtentBytes < (std::numeric_limits<{{ typename_unsigned_bit_length }}>::max() / 8U), {# -#}
+            "This message is too large to be handled by the selected types");
+{%- for field in composite_type.fields_except_padding %}
+    {%- if loop.first %}
+        struct TypeOf
```

Comment:
TypeOf needs docs. Also, this is a really generic name that suggest some greater type system. Perhaps call it `TypeOfService` such that you get `TypeOfService::Response` which is a nice, English sentence "The type of the service response." etc.

Reply:
As long as you think this convention will be stable. We need to start closing the design of the C++ objects so we shouldn't toss in ergonomic things that we aren't confident are going to stand up over time.